src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -808,6 +812,94 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
+        {
+            Debug.Assert(arguments.AttributeSyntaxOpt is not null);
+            if (!IsStatic || MethodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction))
+            {
+                // `UnmanagedCallersOnly` can only be applied to ordinary static methods or local functions.
+                arguments.Diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, arguments.AttributeSyntaxOpt.Location);
+            }
+
+            if (!arguments.Attribute.CommonNamedArguments.IsDefaultOrEmpty)
+            {
+                foreach (var (key, value) in arguments.Attribute.CommonNamedArguments)
+                {
+                    if (key != "CallConvs"
+                        || value.Kind != TypedConstantKind.Array
+                        || value.Values.Any(v => v.Kind != TypedConstantKind.Type))
+                    {
+                        continue;
+                    }
+
+                    foreach (var callConvTypedConstant in value.Values)
+                    {
+                        Debug.Assert(callConvTypedConstant.Kind == TypedConstantKind.Type);
+                        if (!(callConvTypedConstant.ValueInternal is NamedTypeSymbol callConvType)
+                            || !FunctionPointerTypeSymbol.IsCallingConventionModifier(callConvType))
+                        {
+                            // `{0}` is not a valid calling convention type for 'UnmanagedCallersOnly'.
+                            arguments.Diagnostics.Add(ErrorCode.ERR_InvalidUnmanagedCallersOnlyCallConv, arguments.AttributeSyntaxOpt.Location, callConvTypedConstant.ValueInternal ?? "null");
+                        }
+                    }
+                }
+            }
+
+            var (returnTypeSyntax, genericTypeParameters) = SyntaxNode switch
+            {
+                MethodDeclarationSyntax m => (m.ReturnType, m.TypeParameterList),
+                LocalFunctionStatementSyntax l => (l.ReturnType, l.TypeParameterList),
+                _ => default
+            };
+
+            if (returnTypeSyntax is null)
+            {
+                // We already issued an error for these above. We don't immediately bail out so we can error on invalid
+                // calling convention types as well, but erroring on parameter or return types will likely just be noise.
+                Debug.Assert(MethodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction));
+                return;
+            }
+
+            // Methods decorated with `UnmanagedCallersOnly` are required to have all their parameters and return types
+            // be blittable. This means they must be fully unmanaged (not UnmanagedWithGenerics), and it means the method
+            // cannot have type parameters itself, even if those type parameters are constrained to unmanaged.
+
+            if (IsGenericMethod)
```

Comment:
>IsGenericMethod [](start = 16, length = 15)

What if the containing type is generic?

Reply:
I'm actually not sure if that would be supported, I'll clarify that as well.

---
In reply to: [476818894](https://github.com/dotnet/roslyn/pull/47008#discussion_r476818894) [](ancestors = 476818894)